File: JavaScript/Math-Array-String/45.js
// 643. Maximum Average Subarray I

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findMaxAverage = function(nums, k) {
    // Calculate the sum of the first 'k' elements
    let currentSum = 0;
    for (let i = 0; i < k; i++) {
        currentSum += nums[i];
    }

    // Initialize maxSum with the sum of the first window
    let maxSum = currentSum;

    // Slide the window across the array
    for (let i = k; i < nums.length; i++) {
        currentSum = currentSum + nums[i] - nums[i - k];
        maxSum = Math.max(maxSum, currentSum);
    }

    // Return the maximum average
    return maxSum / k;
};

// Example 1
const nums1 = [1, 12, -5, -6, 50, 3];
const k1 = 4;
console.log(findMaxAverage(nums1, k1)); // Output: 12.75

// Example 2
const nums2 = [5];
const k2 = 1;
console.log(findMaxAverage(nums2, k2)); // Output: 5.0

File: Python/Math-Array-String/45.py
# 643. Maximum Average Subarray I

from typing import List

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
                # Calculate the sum of the first 'k' elements
                current_sum = sum(nums[:k])
                max_sum = current_sum

                # Slide the window through the rest of the array
                for i in range(k, len(nums)):
                    current_sum = current_sum + nums[i] - nums[i - k]
                    max_sum = max(max_sum, current_sum)

                # Return the maximum average
                return max_sum / k
    

sol = Solution()
output = sol.findMaxAverage([1,12,-5,-6,50,3], k = 4) # Output = 12.75
print(output)